ons_alpha/private_media/signal_handlers.py

```diff
@@ -0,0 +1,103 @@
+from collections import defaultdict
+
+from django.apps import apps
+from django.contrib.contenttypes.models import ContentType
+from django.db.models import Q, QuerySet
+from wagtail.images.models import AbstractImage
+from wagtail.models import Page, ReferenceIndex
+from wagtail.signals import page_published, page_unpublished
+
+from .models import PrivateMediaCollectionMember
+
+
+def get_private_media_models():
+    return [m for m in apps.get_models() if issubclass(m, PrivateMediaCollectionMember)]
+
+
+def publish_page_media_on_publish(instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Signal handler to be connected to the 'page_published' signal for
+    all page types. It is responsible for identifying any privacy-controlled
+    media used by the page, and ensuring that it is also made public.
+    """
+    ids_by_ctype = defaultdict(list)
+    for ct_id, obj_id in get_all_page_media_ids(instance):
+        ids_by_ctype[ct_id].append(obj_id)
+
+    for ct_id, id_list in ids_by_ctype.items():
+        model_class = ContentType.objects.get_for_id(ct_id).model_class()
+        id_list = [model_class._meta.pk_field.to_python(id) for id in id_list]
+        private_objects_qs = model_class.objects.filter(id__in=id_list, is_private=True).select_related(
+            "collection", "last_public_collection", "last_private_collection"
+        )
+        if issubclass(model_class, AbstractImage):
+            private_objects_qs = private_objects_qs.prefetch_related("renditions")
+        model_class.objects.bulk_make_public(list(private_objects_qs))
+
+
+def unpublish_page_media_on_unpublish(instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Signal handler to be connected to the 'page_unpublished' signal for
+    all page types. It is responsible for identifying any privacy-controlled
+    media used exclusively by the page, and ensuring that it is also made
+    private.
+    """
+    ids_by_ctype = defaultdict(list)
+    for ct_id, obj_id in get_unique_to_page_media_ids(instance):
+        ids_by_ctype[ct_id].append(obj_id)
+
+    for ct_id, id_list in ids_by_ctype.items():
+        model_class = ContentType.objects.get_for_id(ct_id).model_class()
+        id_list = [model_class._meta.pk_field.to_python(id) for id in id_list]
+        public_objects_qs = model_class.objects.filter(id__in=id_list, is_private=False).select_related(
+            "collection", "last_public_collection", "last_private_collection"
+        )
+        if issubclass(model_class, AbstractImage):
+            public_objects_qs = public_objects_qs.prefetch_related("renditions")
```

Comment:
To clarify this will handle renditions and ensure any generated ones will be made private?

Reply:
Yes, `bulk_make_private()` calls each objects `get_privacy_controlled_files()` method, which, for images, includes the source file, and those of any renditions.